releases/ReleaseCompatibility.md

```diff
@@ -0,0 +1,32 @@
+# Releases & Compatibility
+
+To date, we have been somewhat cavalier about making breaking changes in Fairlearn.
+Although we are currently pre-`v1.0.0` and hence without particular commitments to compatibility (as generally understood - see e.g. the [Semantic Versioning Scheme](https://semver.org/)), we should work to reduce the number of breaking changes we make.
+
+## The Problem
+
+At the time of writing, we are about to release `v0.4.6`, which has some substantial breaking changes in the Metrics area of Fairlearn.
+Before it, `v0.4.5` reworked a smaller subset of this functionality, which caused smaller breakages.
+There are also further changes to Metrics planned, which may well cause further breaks (although these should be more minor).
+All this is obviously undesirable from a user standpoint - these look like 'patch' level releases, but are actually breaking their code.
+Concretely, we've had users install Fairlearn with `pip`, and then find themselves unable to run Notebooks from our GitHub project - not because the functionality was missing, but because it had been renamed.
+
+We do not need deprecation policies as elaborate of those of [SciKit-Learn](https://numpy.org/neps/nep-0023-backwards-compatibility.html) or [NumPy](https://numpy.org/neps/nep-0023-backwards-compatibility.html) - indeed, policies such as those would be ridiculous given the size of our code and user bases.
+However, we do need to start to move in that direction, or we will not be able to grow our user base due to chaos in the code.
+
+## Proposed Solution
+
+Starting with `v0.5.0` we should make a commitment that anything which works at `v0.n.m_0` will also work for `v0.n.m` so long as `m >= m_0`.
+However, we do *not* guarantee compatibility between `n` and `n+1` in this scheme (although we would seek to minimise breakage).
+
+We will enforce this using builds which run the notebooks currently in the repository against versions of Fairlearn published on PyPI.
```

Comment:
Aren't we moving away from notebooks?

This also has to do with release workflow. Here's a proposal: 

We could say `0.xx.y+1` is a _patch_ or _bug fix_ release to `0.xx.y`.
The `master` branch is `0.xx+1.dev0`, and therefore changes on master can be breaking backward compatibility.
Whenever we want to push a patch release out, we _backport_ the PRs which we'd like to include in `0.xx.y+1` from the master, into the branch for the release.

This also implies that there is a branch for at least each major release, and minor releases are tags on at different times on that branch. The different versions of the documentation are also generated from those branches.

In my experience checking for backward compatibility in reviews is not hard, and if we get used to caring about it, then it'll be fine. In terms of testing, if we really want to test against a different release, we could run the tests against a different release, instead of running the notebooks.

Reply:
> moving away from notebooks?

Yes.... I should at the very least rephrase to be about running things from the `examples` directory.

However, your point about just running the tests is also good. I may have been overly concerned about saving on CPU cycles.

@romanlutz and @MiroDudik , what do you think about the release and branching strategy @adrinjalali is proposing? It looks reasonable to me, but would be more overhead over what we're currently doing.

Reply:
I've updated the proposal, making it a lot closer to the procedure described by @adrinjalali above.

Reply:
Re notebooks - yes! It's just not done yet. Perhaps we should prioritize it.

We're pre-1.0 so we can currently only mess with 0.*.*. The only way to distinguish major releases from minor ones is therefore this last level (patch level). Otherwise we'll jump from 0.5.0 to 0.6.0 to 0.7.0 rapidly. It's really impossible to tell for users whether 0.7.0 is a major change or not at that point, since everything is flat. So fort that reason I'm not a huge fan.

I agree with testing against releases, not just running notebooks. That was just a quick band aid.

Reply:
"Not a huge fan" of what exactly - there are a few things going on here. I think you're concerned that our breaking changes are going to result in us going through minor versions quickly?

Reply:
I think I was confusing myself a little bit. I think it's important to have a consistent approach and this is one such approach. No objections.